dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+
+instantiate
+  :: Substitutable f
+  => Name        -- name
+  -> Free f Var  -- arg
+  -> Free f Var  -- an expression to substitute in
+  -> Free f Var
+instantiate name arg = subst $ Bound name `Map.singleton` arg
+
+capture :: Refreshes m => Name -> m (Subst f)
+capture name = do
+  name' <- refresh name
+  return $ FreeVar name `Map.singleton` Bound name'
+```
+
+Notice, that iе doesn't mention the structure of `AST_` anywhere, which means we have succesfully abstracted it out.
+
+Another thing that will help us is an occurence checker.
+```haskell
+freeVars :: Free f Var -> Set Name
+freeVars = foldMap \case
+  FreeVar n -> Set.singleton n
+  _         -> mempty
+
+occurs :: Name -> Free f Var -> Bool
+occurs n p = n `Set.member` freeVars p
+```
+
+I also recommend writing pattern synonyms for each constructor, so you can mostly forget that `Free` even was there.
+```haskell
+pattern Lam arg  body     = Free (Lam_ arg  body)
+pattern Let name decl ctx = Free (Lam_ name decl ctx)
+```
+
+_For merging `Subst`-s_, apply one of them to all values in other one first.
+
+## Scopes
+
+We still can't construct programs correctly, let clean this problem up.
+
+I suggest using smart constructor to build `AST`, they also perform capture when needed:
+
+```haskell
+lam :: Refreshes m => Name -> AST -> AST -> m AST
+lam name ty body = do
+  s <- capture name
+  return $ Lam (substName s name) ty (subst s body)
+
+let_ :: Refreshes m => Name -> AST -> AST -> AST -> m AST
+let_ name ty decl body = do
+  s <- capture name
+  return $ Let (Val (substName s name) ty decl) (subst s body)
+```
+We capture the name at binder position as well, because our renaming strategy is not structural, unlike the one used by de Brujin indexing. We have to do it, because otherwise there's no way to tell which of the `a#172` is the `a` defined there. And wil _will_ get big indices.
+
+With this approach, we can carefully control which part of the program is closed over which set of names. For instance, I've left `let` statement non-recursive in this example, and in both smart constructors type cannot depend on the argument.
+
+The interesting part is capturing variables in case alternatives.
+```haskell
+case ... of
+  | pat -> expr  -- those ones
+```
+My solution was to treat them like lambdas - by substituting both pattern and expression with variables captured in the pattern.
+
+Using that, we can return scope-correct programs right off the parser.
+
+_Pro tip learned after a debug session_: Refreshing names of records field declarations in record literal is a bad idea.
+
+## Unification
+
+We can construct the program. Before we're doing inference, we need an auxillary mechanism. It whill compare two types and either find a set of substitutions to convert both to the same type, or bail out with error if that isn't possible. This mechanism is called unification.
+
+Usially, in Hindley-Milner inference, the type of unifier is
+
```

Comment:
It's worth referring to some paper on Hindley-Milner inference to make the post more self-contained.